File: ruby/lib/ter/adapters/dotenv.rb
# frozen_string_literal: true

module Ter
  module Adapters
    # DotEnv file parsing and generation
    class DotEnv
      # Parse .env file content
      def self.parse(content)
        result = {}
        current_key = nil
        current_value = ''
        in_multiline = false

        lines = content.split("\n")

        lines.each do |line|
          # Handle multiline heredoc
          if in_multiline
            if line.strip == 'EOF'
              result[current_key] = current_value.rstrip
              in_multiline = false
              current_value = ''
            else
              current_value += line + "\n"
            end
            next
          end

          # Handle line continuation
          if current_key && line.end_with?('\\')
            current_value += line[0..-2] + "\n"
            next
          end

          if current_key && !line.start_with?(/\s*#/) && !line.include?('=')
            current_value += line + "\n"
            next
          end

          # Store previous key-value
          if current_key && current_value
            result[current_key] = current_value.rstrip
            current_key = nil
            current_value = ''
          end

          # Skip comments and empty lines
          next if line.strip.empty? || line.strip.start_with?('#')

          # Parse new key-value
          if line.include?('=')
            key, value = line.split('=', 2)
            key = key.strip

            # Check for heredoc
            if value.strip.start_with?('<<')
              in_multiline = true
              current_key = key
              current_value = ''
            else
              # Remove quotes and handle escapes
              value = unescape_value(value.strip)
              result[key] = value
            end
          end
        end

        # Store final key-value
        result[current_key] = current_value.rstrip if current_key && current_value

        result
      end

      # Generate .env file content
      def self.generate(values)
        lines = []

        values.each do |key, value|
          lines << "#{key}=#{escape_value(value)}"
        end

        lines.join("\n")
      end

      # Escape a value for .env format
      def self.escape_value(value)
        return '""' if value.nil? || value.empty?

        # Use quotes for values with special characters
        if value.include?("\n") || value.include?('"') || value.include?('\\') || value.include?('$')
          escaped = value.gsub('\\', '\\\\').gsub('"', '\\"').gsub('$', '\$')
          "\"#{escaped}\""
        else
          value
        end
      end

      # Unescape a value from .env format
      def self.unescape_value(value)
        return '' if value.nil? || value.empty?

        # Handle quoted values
        if (value.start_with?('"') && value.end_with?('"')) ||
           (value.start_with?("'") && value.end_with?("'"))
          # Remove quotes
          inner = value[1..-2]
          # Unescape
          inner.gsub('\\n', "\n").gsub('\\t', "\t").gsub('\\\\', '\\').gsub('\\"', '"')
        else
          value
        end
      end

      # Load from file
      def self.load_file(path)
        raise "File not found: #{path}" unless File.exist?(path)
        content = File.read(path)
        parse(content)
      end

      # Save to file
      def self.save_file(path, values)
        content = generate(values)
        File.write(path, content)
      end
    end
  end
end

File: python/ter/adapters/dotenv.py
"""DotEnv file parsing and generation"""

from typing import Dict, List, Tuple, Optional
from dataclasses import dataclass
import re


@dataclass
class ParseError:
    """Parse error with location"""
    line: int
    message: str


@dataclass
class ParseResult:
    """Result of parsing .env content"""
    valid: bool
    values: Dict[str, str]
    errors: List[ParseError]


def parse_dotenv(content: str) -> ParseResult:
    """Parse .env file content"""
    lines = content.split('\n')
    values: Dict[str, str] = {}
    errors: List[ParseError] = []
    
    for line_num, line in enumerate(lines, 1):
        line = line.strip()
        
        # Skip empty lines and comments
        if not line or line.startswith('#'):
            continue
        
        # Find = separator
        if '=' not in line:
            errors.append(ParseError(line_num, "Missing = separator"))
            continue
        
        key, _, raw_value = line.partition('=')
        key = key.strip()
        
        # Validate key
        if not key or not re.match(r'^[A-Za-z_][A-Za-z0-9_]*$', key):
            errors.append(ParseError(line_num, f"Invalid key name: {key}"))
            continue
        
        # Parse value
        raw_value = raw_value.strip()
        
        # Handle quoted values
        if raw_value.startswith('"') and raw_value.endswith('"'):
            # Double quoted - process escape sequences
            value = raw_value[1:-1]
            value = value.replace('\\n', '\n')
            value = value.replace('\\t', '\t')
            value = value.replace('\\r', '\r')
            value = value.replace('\\"', '"')
            value = value.replace('\\\\', '\\')
        elif raw_value.startswith("'") and raw_value.endswith("'"):
            # Single quoted - no escape processing
            value = raw_value[1:-1]
        else:
            # Unquoted
            value = raw_value
        
        values[key] = value
    
    return ParseResult(valid=len(errors) == 0, values=values, errors=errors)


def generate_dotenv(values: Dict[str, str], header: Optional[str] = None) -> str:
    """Generate .env content from values"""
    lines: List[str] = []
    
    if header:
        lines.append(f"# {header}")
        lines.append("")
    
    for key, value in values.items():
        # Determine if quoting is needed
        if needs_quoting(value):
            # Escape special characters
            escaped = value.replace('\\', '\\\\')
            escaped = escaped.replace('\n', '\\n')
            escaped = escaped.replace('\t', '\\t')
            escaped = escaped.replace('\r', '\\r')
            escaped = escaped.replace('"', '\\"')
            lines.append(f'{key}="{escaped}"')
        else:
            lines.append(f'{key}={value}')
    
    return '\n'.join(lines)


def needs_quoting(value: str) -> bool:
    """Check if value needs quoting"""
    if not value:
        return True
    if ' ' in value:
        return True
    if any(c in value for c in ['=', '"', "'", '#', '\n', '\t']):
        return True
    return False


def merge_dotenv(*dicts: Dict[str, str]) -> Dict[str, str]:
    """Merge multiple environment dictionaries"""
    result: Dict[str, str] = {}
    for d in dicts:
        result.update(d)
    return result


def diff_dotenv(before: Dict[str, str], after: Dict[str, str]) -> Dict[str, any]:
    """Diff two environment dictionaries"""
    added = {}
    removed = {}
    modified = {}
    
    # Added and modified
    for key, value in after.items():
        if key not in before:
            added[key] = value
        elif before[key] != value:
            modified[key] = {
                "before": before[key],
                "after": value
            }
    
    # Removed
    for key in before:
        if key not in after:
            removed[key] = before[key]
    
    return {
        "added": added,
        "removed": removed,
        "modified": modified
    }


def validate_dotenv(content: str) -> bool:
    """Check if .env content is valid"""
    result = parse_dotenv(content)
    return result.valid
